source/fab/steps/grab.py

```diff
@@ -28,25 +32,33 @@
 
 class GrabSourceBase(Step, ABC):
     """
-    Base class for grab steps. All grab steps require a source and a folder in which to put it.
+    Base class for source grab steps.
 
     Unlike most steps, grab steps don't need to read or create artefact collections.
 
+    At runtime, when the build config is available, the destination folder is stored into `self._dst`.
+
     """
-    def __init__(self, src: str, dst: Optional[str] = None, name=None):
+    def __init__(self, src: str, dst: Optional[str] = None, revision=None, name=None):
         """
         :param src:
-            The source location to grab. The nature of this parameter is depends on the subclass.
+            The source url to grab.
         :param dst:
-            The name of a sub folder, in the project workspace, in which to put the source.
+            The name of a sub folder in the project workspace's source folder, in which to put this source.
             If not specified, the code is copied into the root of the source folder.
+        :param revision:
+            E.g 'vn6.3' or 36615
         :param name:
-            Human friendly name for logger output, with sensible default.
+            Human friendly name for logger output, with a sensible default.
 
         """
-        super().__init__(name=name or f'{self.__class__.__name__} {dst or src}'.strip())
+        super().__init__(name=name or f'{self.__class__.__name__} {src} {revision}'.strip())
         self.src: str = src
         self.dst_label: str = dst or ''
+        self.revision = revision
+
+        # runtime
+        self._dst: Optional[Path] = None
```

Comment:
Your typing problems may well be down to the fact that this member is declared `Optional` but it is not fenced with `if self._dst is not None` when used.

However there is a pungent code smell here. It looks like you're using a nominally private member variable to communicate between child and parent classes.

At the very least some sort of accessor function should be used rather than direct access. A `get_destination` method or a `@property` method.

Reply:
I'd argue that there's no such thing as communication between a parent and child class but that's moot because the property suggestion is nicer anyway!

Reply:
Ah, I couldn't do a property because it depends on another runtime attribute, the config.

Interestingly, that config is also made available using the same _"runtime attribute pattern"_ (made up term) via `self._config` so child processes can read it.

The fortran compiler also uses two runtime-only attributes called `self._stage` and `self._mod_hashes`, and the two language parsers both have one.

Perhaps we should have an OO chat about this?

Reply:
We should definitely have a discussion about this but probably not in this change. I've raised #210 as a venue.